### PR TITLE
Rholang: Add support for methods

### PR DESCRIPTION
**Pull request by guardbotmk3**
Adds support for methods in the normalizer and the Reducer.

### CODE REVIEW

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 1]
```diff
@@ -514,5 +550,30 @@ object Reduce {
...
+              Task raiseError new Error("Error: nth expects 1 argument")
+            for {
+              nth <- evalToInt(args(0))(env)
+              v   <- evalSingleExpr(p)(env)
+              result <- v.exprInstance match {
+                         case EListBody(EList(ps, _, _, _))   => localNth(ps, nth)
+                         case ETupleBody(ETuple(ps, _, _, _)) => localNth(ps, nth)
+                         case _ =>
+                           Task raiseError new Error(
+                             "Error: nth applied to something that wasn't a list or tuple.")
```
Comment (KentShikama) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Are there plans to take more than either a list or tuple? It seems like just taking either would suffice.

Comment (guardbotmk3) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: I did not have plans to make the nth method specifically handle anything but tuples and lists.
Other methods may support additional kinds of data.
For instance, Mike has plans for us to implement confinement as a method on arbitrary processes.

Comment (KentShikama) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Got it

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 2]
```diff
@@ -381,17 +376,54 @@ object Reduce {
...
           for {
             evaledPs  <- el.ps.toList.traverse(expr => evalExpr(expr)(env))
             updatedPs = evaledPs.map(updateLocallyFree)
           } yield updateLocallyFree(EList(updatedPs, el.freeCount, el.locallyFree, el.wildcard))
         }
+        case EMethodBody(EMethod(method, target, arguments, _, _, _)) => {
+          for {
+            evaledTarget <- evalExpr(target.get)
+            evaledArgs   <- arguments.toList.traverse(expr => evalExpr(expr)(env))
+            methodLookup = methodTable.get(method)
```
[the remark this answers is not in the document]
Comment (guardbotmk3) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Done.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 3]
```diff
@@ -381,17 +376,54 @@ object Reduce {
...
+        case EVarBody(EVar(v)) =>
+          for {
+            p       <- eval(v.get)
+            evaledP <- evalExpr(p)
+          } yield evaledP
+        case EMethodBody(EMethod(method, target, arguments, _, _, _)) => {
+          for {
+            evaledTarget <- evalExpr(target.get)
+            evaledArgs   <- arguments.toList.traverse(expr => evalExpr(expr)(env))
+            methodLookup = methodTable.get(method)
```
[the remark this answers is not in the document]
Comment (guardbotmk3) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Done.

**Comment by guardbotmk3** [comment]
OK, all comments addressed, sorting added, we should be good to go.

**Pull request merged**